// !LANGUAGE: +ProperCheckAnnotationsTargetInTypeUsePositions -ClassTypeParameterAnnotations

annotation class A1
annotation class A2(val some: Int = 12)

@Target(AnnotationTarget.TYPE)
annotation class TA1

@Target(AnnotationTarget.TYPE_PARAMETER)
annotation class TA2(val some: Int = 12)

class TopLevelClass<<!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A1<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A2(3)<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A2<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A1(12)<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A2("Test")<!> T> {
    class InnerClass<<!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A1<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A2(3)<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A2<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A1(12)<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A2("Test")<!> T> {
        fun test() {
            class InFun<<!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A1<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A2(3)<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A2<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A1(12)<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@A2("Test")<!> T>
        }
    }
}

class TTopLevelClass<<!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA1<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA2(3)<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA2<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA1(12)<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA2("Test")<!> T> {
    class TInnerClass<<!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA1<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA2(3)<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA2<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA1(12)<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA2("Test")<!> T> {
        fun test() {
            class TInFun<<!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA1<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA2(3)<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA2<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA1(12)<!> <!UNSUPPORTED, WRONG_ANNOTATION_TARGET!>@TA2("Test")<!> T>
        }
    }
}
